<a href="https://colab.research.google.com/github/JuCaRiCo/202602-AnaliticaDescriptivaPredictiva/blob/main/Actividad_Calificada_07_Datos_Duplicados_271760.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Juan Carlos Rivera Cordova| |
| **Matrícula** 271760 | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:

#Total de records existentes
totalRecords = len(hoteles)

#Total de records duplicados usando 'first'
duplicatedFirst = hoteles.duplicated(keep='first').sum()
print("Duplicados (keep='first'): ", duplicatedFirst)

#Total de records duplicados usando 'last'
duplicatedLast = hoteles.duplicated(keep='last').sum()
print("Duplicados (keep='last'): ", duplicatedLast)

#Total de records duplicados usando 'False'
duplicatedFalse = hoteles.duplicated(keep=False).sum()
print('Duplicados (keep=False): ', duplicatedFalse)
print('')
print("Explicación de significados:")
print("keep='first' (", duplicatedFirst, "): hace referencia a los records repetidos conservando el primer record ")
print("keep='last' (", duplicatedLast, "): hace referencia a los records repetidos conservando el ultimo record")
print("keep=False (", duplicatedFalse, "): hace referencia a los records que no son unicos")


Duplicados (keep='first'):  31994
Duplicados (keep='last'):  31994
Duplicados (keep=False):  40165

Explicación de significados:
keep='first' ( 31994 ): hace referencia a los records repetidos conservando el primer record 
keep='last' ( 31994 ): hace referencia a los records repetidos conservando el ultimo record
keep=False ( 40165 ): hace referencia a los records que no son unicos


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# sacamos el porcentaje dividiendo duplicados entre total de records, y multiplicamos por 100,
# haciendo round a 2 decimales
porcentajeSobrantes = round(duplicatedFirst / totalRecords * 100, 2)
print('Porcentaje de records sobrantes respecto al total: ', porcentajeSobrantes, '%')

# sacamos los duplicados y lo usamos como subconjunto de hoteles y aplicamos drop_duplicates
# para remover el total duplicados y dejar solo un registro de los duplicados
gruposRepetidos = len(hoteles[hoteles.duplicated(keep=False)].drop_duplicates())
print('Numero de grupos repetidos: ', gruposRepetidos)


Porcentaje de records sobrantes respecto al total:  26.8 %
Numero de grupos repetidos:  8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
hotelBkp = hoteles.copy()

# se crea una columna para guardar el valor booleando para saber si es duplicado
hotelBkp['es_duplicado'] = hoteles.duplicated(keep='first')

#Resumen Total por Hotel
totalPorHotel = hotelBkp.groupby('hotel')['es_duplicado'].count()
#Resumen de Sobrantes por hotel
sobrantesPorHotel = hotelBkp.groupby('hotel')['es_duplicado'].sum()
#Porcentaje sobrantes sobre total por hotel
porcentajePorHotel = round(sobrantesPorHotel / totalPorHotel * 100, 2)

print('Total Por Hotel:\n',totalPorHotel)
print('\nSobrantes Por Hotel:\n',sobrantesPorHotel)
print('\nPorcentaje Sobrantes Por Hotel:\n',porcentajePorHotel)

Total Por Hotel:
 hotel
City Hotel      79330
Resort Hotel    40060
Name: es_duplicado, dtype: int64

Sobrantes Por Hotel:
 hotel
City Hotel      25902
Resort Hotel     6092
Name: es_duplicado, dtype: int64

Porcentaje Sobrantes Por Hotel:
 hotel
City Hotel      32.65
Resort Hotel    15.21
Name: es_duplicado, dtype: float64


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema es alto, dado que tenemos un 26.8% de duplicidad, auqnue si lo desglosamos por hotel, se ve mas afectado considerablemente "City Hotel" con un 32.65%, mas del doble del "Resort Hotel" que tiene un 15.21%

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
columnasParaMostrar = [
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adults',
    'country',
    'market_segment',
    'deposit_type',
    'adr',
    'reservation_status',
]

recordsDuplicados = hoteles[hoteles.duplicated(keep=False)]

#la eleccion de tomar columnas de fecha es para poderlas ordenar cronologicamente
columnasParaOrden = [
    'hotel',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'lead_time',
]

recordsDuplicadosOrdenados = recordsDuplicados.sort_values(by=columnasParaOrden)
recordsDuplicadosOrdenados.head(12)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
40574,City Hotel,1,2,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-30
40580,City Hotel,1,2,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-30
40588,City Hotel,1,2,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-30
40562,City Hotel,0,117,2015,August,31,1,2,1,1,...,No Deposit,17.0,NaN,0,Transient-Party,72.5,0,0,Check-Out,2015-08-04
40563,City Hotel,1,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-28
40564,City Hotel,1,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-28
40566,City Hotel,0,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Check-Out,2015-08-04
40567,City Hotel,0,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Check-Out,2015-08-04
40568,City Hotel,1,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-28
40569,City Hotel,1,117,2015,August,31,1,2,1,2,...,No Deposit,17.0,NaN,0,Transient-Party,94.5,0,0,Canceled,2015-07-28


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [56]:
conteoCombinaciones = hoteles.value_counts(dropna=False)

# Obetenemos el conteo por grupo con value_counts
tamanosGrupos = conteoCombinaciones.value_counts().head(10)
print('los 10 grupos con mayor frecuencia: \n',tamanosGrupos)


print('--------------------------------------------------')
# sacamos el max para tener el de maor tamano
grupoMayorTamano = conteoCombinaciones.max()

print('Grupo de Mayor Tamano: ', grupoMayorTamano)
print('--------------------------------------------------')
# le damos un mejor formato con pd.Series, usando las columnas
print(pd.Series(conteoCombinaciones.index[0], index=hoteles.columns))



los 10 grupos con mayor frecuencia: 
 count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64
--------------------------------------------------
Grupo de Mayor Tamano:  180
--------------------------------------------------
hotel                             City Hotel
is_canceled                                1
lead_time                                277
arrival_date_year                       2016
arrival_date_month                  November
arrival_date_week_number                  46
arrival_date_day_of_month                  7
stays_in_weekend_nights                    1
stays_in_week_nights                       2
adults                                     2
children                                 0.0
babies                                     0
meal                                      BB
country                                  PRT
market_segment                        Groups
distr

\### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [57]:
# filtramos con keep=False
todosLosDuplicados = hoteles[hoteles.duplicated(keep=False)]

# Distribución de market_segment
marketSegmentTotal = hoteles['market_segment'].value_counts(normalize=True) * 100
marketSegmentDuplicados = todosLosDuplicados['market_segment'].value_counts(normalize=True) * 100

distMarketSegment = pd.DataFrame(
    {'Dataset Completo %': marketSegmentTotal, 'Recors Duplicadas %': marketSegmentDuplicados}
)

print('Distribucion Market Segment: \n', distMarketSegment.round(2))

print('\n--------------------------------------------------')
# Distribución de deposit_type
depositTypeTotal = hoteles['deposit_type'].value_counts(normalize=True) * 100
depositTypeDuplicados = todosLosDuplicados['deposit_type'].value_counts(normalize=True) * 100

distDepositType = pd.DataFrame(
    {'Dataset Completo %': depositTypeTotal, 'Records Duplicados %': depositTypeDuplicados}
)

print('Distribucion Deposit Type: \n', distDepositType.round(2))

print('\n--------------------------------------------------')

#tasa de cancelacion
tasaCancelTotal = hoteles['is_canceled'].mean() * 100
tasaCancelDuplicadoss = todosLosDuplicados['is_canceled'].mean() * 100

distCancelacion = pd.DataFrame(
    {'Dataset Completo %': [tasaCancelTotal],
        'Records Duplicadas %': [tasaCancelDuplicadoss],
    },
    index=['Tasa de Cancelacion'],
)

print('Distribucion Cancelacion: \n', distCancelacion.round(2))

Distribucion Market Segment: 
                 Dataset Completo %  Recors Duplicadas %
market_segment                                         
Aviation                      0.20                 0.05
Complementary                 0.62                 0.19
Corporate                     4.44                 3.67
Direct                       10.56                 3.50
Groups                       16.59                41.61
Offline TA/TO                20.29                30.40
Online TA                    47.30                20.59
Undefined                     0.00                  NaN

--------------------------------------------------
Distribucion Deposit Type: 
               Dataset Completo %  Records Duplicados %
deposit_type                                          
No Deposit                 87.65                 64.17
Non Refund                 12.22                 35.64
Refundable                  0.14                  0.19

--------------------------------------------------
D

### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

Respuesta A)
- Para el segmento Groups,asumiendo reservas reales, tenemos que para el dataset completo representa un 16.59%, mientras que el apartado de duplicados tenemos 41.61%. Esto esta dado porque las reservaciones en grupos generalmente son identicas para todo el grupo.
- Para Reservas que podrian ser errores o duplicados por alguna razon, tenemos que en las canceladas muestra una duplicidad de 58.39%, contra el 37.04% del dataset completo, mostrando que efectivamente podrian ser errores o duplicaciones por sistema al intentar cancelar y tener algun fallo en el intento.Lo mismo pasa con Non Refund, la diferencia es mucha, mas considerando que son mas rigidos al hacer devoluciones, como para tener un "error" tan grande.

Respuesta B)
Lo que podria entender, es que son reservaciones realizadas por una persona y no por un agente o compania, lo cual podria hacer que marque como duplicados records que no lo son

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [58]:
print('Numero de Records previo a limpiar el DS: ', len(hoteles))
hotelesLimpio = hoteles.drop_duplicates(keep='first', ignore_index=True)
print('Numero de Records despues de limpiar el DS: ', len(hotelesLimpio))



Numero de Records previo a limpiar el DS:  119390
Numero de Records despues de limpiar el DS:  87396


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [59]:
print('Numero de Reservas: ')
cantOriginal = hoteles.groupby('hotel').size().reset_index()
cantLimpio = hotelesLimpio.groupby('hotel').size().reset_index()

print('Cantidad Original: ', cantOriginal)
print('\nCantidad Limpia: ', cantLimpio)

print('\n---------------------------------------------------------------')
print('Tasa de Cancelacion (media de is_canceled)')
mediasOriginal = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()
mediasLimpio = hotelesLimpio.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()

print('Media Original: ', mediasOriginal)
print('\nMedia Limpia: ', mediasLimpio)

print('\n---------------------------------------------------------------')
print('Tarifa diaria promedio')
resumen_original = pd.merge(cantOriginal, mediasOriginal, on='hotel')
resumen_limpio = pd.merge(cantLimpio, mediasLimpio, on='hotel')

print('Resumen Original: ', resumen_original)
print('\nResumen Limpia: ', resumen_limpio)


Numero de Reservas: 
Cantidad Original:            hotel      0
0    City Hotel  79330
1  Resort Hotel  40060

Cantidad Limpia:            hotel      0
0    City Hotel  53428
1  Resort Hotel  33968

---------------------------------------------------------------
Tasa de Cancelacion (media de is_canceled)
Media Original:            hotel  is_canceled         adr
0    City Hotel     0.417270  105.304465
1  Resort Hotel     0.277634   94.952930

Media Limpia:            hotel  is_canceled         adr
0    City Hotel     0.300386  110.985944
1  Resort Hotel     0.234809   99.025346

---------------------------------------------------------------
Tarifa diaria promedio
Resumen Original:            hotel      0  is_canceled         adr
0    City Hotel  79330     0.417270  105.304465
1  Resort Hotel  40060     0.277634   94.952930

Resumen Limpia:            hotel      0  is_canceled         adr
0    City Hotel  53428     0.300386  110.985944
1  Resort Hotel  33968     0.234809   99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [60]:
# Creamos copias explicitas para agregar la columna 'version' a cada resumen
resumenOriginalV = resumen_original.copy()
resumenOriginalV['version'] = 'original'

resumenLimpioV = resumen_limpio.copy()
resumenLimpioV['version'] = 'sin duplicados'

print('\n---------------------------------------------------------------')
comparacion = pd.concat([resumenOriginalV, resumenLimpioV], ignore_index=True)

# Ordenamos por hotel, version
comparacion = comparacion.sort_values(by=['hotel', 'version']).reset_index(drop=True)

comparacion




---------------------------------------------------------------


,hotel,0,is_canceled,adr,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

Respuesta A)
Para City Hotel, is_canceled cambia drasticamente, con una diferencia de al rededor de 11 puntos porcentuales, Resort Hotel no presenta una diferencia tan grande, ya que es solo de aproximadamente el 4%, por otro lado ADR no varia mucho para ambas, ambas con al rededor de 5 puntos. Con relacion a la actividad dos, se ve claramente la diferencia tambien entre las cancelaciones, teniendo una coincidencia con estos resultados.

Respuesta B)
- Entrenamiento: habria una gran discrepancia o sobreponderancia con respecto a Groups y Cancelados, dado que los duplicados son demasiados
- Evaluacion: Al igual que en el entrenamiento tendriamos un desempeño muy distorcionado

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [62]:
# Se hace el conteo sobre hoteles limpios
duplicadosRestantes = hoteles_limpio.duplicated().sum()
print('Diplicados Restantes: ', duplicadosRestantes)

# Validando Indices
indiceUnico = hoteles_limpio.index.is_unique
print('Indice es unico: ', indiceUnico)

# Verificacion del rango consecutivo del índice de 0 a len(hoteles_limpio)
rangoEsperado = pd.RangeIndex(start=0, stop=len(hoteles_limpio))
indiceEsCnsecutivo = hoteles_limpio.index.equals(rangoEsperado)

print('Indice Es Consecutivo: ', indiceEsCnsecutivo)

if duplicadosRestantes == 0 and indiceUnico and indiceEsCnsecutivo:
    print("\nVerificacion exitosa!")
else:
    print("\nFallo la erificacion.")

Diplicados Restantes:  0
Indice es unico:  True
Indice Es Consecutivo:  True

Verificacion exitosa!


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [71]:
# Obtenemos la lista de columnas excluyendo 'reservation_status_date'
colsSinFecha = [col for col in hoteles_limpio.columns if col != 'reservation_status_date']

# Identificamos los duplicados (keep='first')
duplicadosParcialesSobrantes = hoteles_limpio.duplicated(subset=colsSinFecha, keep='first').sum()
print('Duplicados Parciales Sobrantes: ', duplicadosParcialesSobrantes)

# Filtramos las filas involucradas en este tipo de duplicado (keep=False)
recordsParciales = hoteles_limpio[hoteles_limpio.duplicated(subset=colsSinFecha, keep=False)]
print('Filtrado de duplicados: \n',recordsParciales)

colsParaMostrar = [
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adr',
    'reservation_status',
    'reservation_status_date',
]
colsParaOrden = [
    'hotel',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'lead_time',
]

parcialesOrdenados = recordsParciales.sort_values(by=colsParaOrden)
print('\nRecords con duplicacion parcial\n', parcialesOrdenados[colsParaMostrar].head(10).to_string())

# Distribucion de reservation_status en estas filas
distStatus = parcialesOrdenados['reservation_status'].value_counts(normalize=True) * 100
print('\nDistribución de reservation_status en duplicados parciales: \n', distStatus.round(2))

Duplicados Parciales Sobrantes:  245
Filtrado de duplicados: 
               hotel  is_canceled  lead_time  arrival_date_year  \
1911   Resort Hotel            1         81               2015   
1913   Resort Hotel            1         81               2015   
2735   Resort Hotel            1         41               2015   
2736   Resort Hotel            1         41               2015   
2794   Resort Hotel            1         36               2015   
...             ...          ...        ...                ...   
55682    City Hotel            1         32               2015   
55781    City Hotel            1         24               2015   
56010    City Hotel            1         24               2015   
56448    City Hotel            1         21               2016   
56789    City Hotel            1         21               2016   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
1911           September                        40             

**Respuesta 4.2:**

El resultado muestra que el 100 corresponde a reservaciones canceladas, y viendo el status date, todos los records estan duplicados salvo el status date, pareciera que se ingreso la cancelacion dos veces para todos esos records, y si, deberian borrarse esos registros, manteniendo el ultimo registro con keep='last'

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Registro de Decisión de Limpieza

Dataset: Hotel Booking Demand

Módulo: Tratamiento de Duplicados

1. Criterio de Duplicado Utilizado
- Duplicidad exacta, se eliminaron records donde todas sus columnas eran duplicadas exactamente.

2. Regla de Conservación
- Parametro Utilizado: keep='first'
- Relevancia: Si bien no hay diferencia entre 'first' y 'last', dado que todos los records eran identicos en todas sus columnas, se prefiere 'first' para hacer el reinidice secuencial

3. Resultado Cuantitativo
- Records Totales Originales: 119,390
- Records Eliminados (Duplicados Exactos): 31,994 (26.79%)
- Records Finales: 87,396
4. Supuestos y Limitaciones
- Supuestos: Se asume que los registros identicos son errores del sistema y no reservaciones legitimas
- Limitaciones: Es posible que al eliminar reservas duplicadas exactas, hayan sido reservaciones reales de reservaciones por grupo, con respecto a los duplicados parciales que corresponden a cancelaciones, quedaron pendientes de remover, si es que se decide conservar el ultimo registro asumiendo que se registro dos veces la cancelacion.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. Podria ser en caso de alguna auditoria, ya que muchas de las reservaciones que son exactamente iguales podrian pertenecer a reservaciones por grupo, que suelen llevar mismas fechas, etc.
2. Tener un ID unico para las reservas como un booking_id, o un reservation_id que sea una llave primaria, otra opcion seria tener un customer_id, para saber si son duplicadas o en realidad pertenecen a otro cliente, adicionalmente podria tenerse una columna para el created_date, con timestamp, para saber la hora exacta y poder diferenciar por esa columna si corresponde a duplicidad o un registro diferente.
3. Como se vio en el caso de duplicidad parcial, que pertenecias an cancelados, podrian ser registros que fueron hecho en dos intentos por cancelar, y estos serian ignorados por esta tecnica de validacion.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.